core/vm/gas_table.go

```diff
@@ -402,6 +402,15 @@ func gasCall(evm *EVM, contract *Contract, stack *Stack, mem *Memory, memorySize
 	if gas, overflow = math.SafeAdd(gas, evm.callGasTemp); overflow {
 		return 0, ErrGasUintOverflow
 	}
+	if evm.chainRules.IsEIP4762 {
```

Comment:
I think it's wrong here.

In EIp158, we introduce a rule that at most 63/64 gas could be used in a inner call, therefore the gas for CALL opcode should be fully computed, and then 1/64 remaining gas will be deducted for this rule.

However in this pull request, the verkle specific gas is deducted after this rule. It's a behavioral change.

suggested change:

```diff
-       if transfersValue && !evm.chainRules.IsEIP4762 {
+       if transfersValue {
                gas += params.CallValueTransferGas
        }
        memoryGas, err := memoryGasCost(mem, memorySize)
@@ -394,7 +394,14 @@ func gasCall(evm *EVM, contract *Contract, stack *Stack, mem *Memory, memorySize
        if gas, overflow = math.SafeAdd(gas, memoryGas); overflow {
                return 0, ErrGasUintOverflow
        }
-
+       if evm.chainRules.IsEIP4762 && transfersValue {
+               // Deduct the gas cost for value transfers according to the legacy rules.
+               gas -= params.CallValueTransferGas
+               gas, overflow = math.SafeAdd(gas, evm.AccessEvents.ValueTransferGas(contract.Address().Bytes()[:], address.Bytes()[:]))
+               if overflow {
+                       return 0, ErrGasUintOverflow
+               }
+       }
        evm.callGasTemp, err = callGas(evm.chainRules.IsEIP150, contract.Gas, gas, stack.Back(0))
        if err != nil {
                return 0, err
@@ -402,15 +409,6 @@ func gasCall(evm *EVM, contract *Contract, stack *Stack, mem *Memory, memorySize
        if gas, overflow = math.SafeAdd(gas, evm.callGasTemp); overflow {
                return 0, ErrGasUintOverflow
        }
-       if evm.chainRules.IsEIP4762 {
-               if transfersValue {
-                       gas, overflow = math.SafeAdd(gas, evm.AccessEvents.ValueTransferGas(contract.Address().Bytes()[:], address.Bytes()[:]))
-                       if overflow {
-                               return 0, ErrGasUintOverflow
-                       }
-               }
-       }

```

Reply:
so for the part of the sending value, this is the correct behavior:

![image](https://github.com/ethereum/go-ethereum/assets/3272758/6299c528-295d-4180-a1a1-dbaecff6da9f)

Reply:
I had to think about the other part a little, but I am now convinced that you are correct.